README.md
# google-map-styer

Vanilla javascript plugin for changing style/color on Google Map.

## Usage

1. Set HTML
```
<div class="gmap">
<div class="gmap-inner"></div>
</div>
```
- each needs to have separated class

2. Set CSS
Width & height is defined in CSS.
Set responsive option true if the width / height is flexible.

```
.gmap{
	width:500px;
	height:500px;
}
.gmap-inner{
	width:500px;
	height:500px;
}

```

3. Include scripts  
You'll need to get Google Map API key.

```
<!-- include google map main js-->
<script id="gmap" src="http://maps.google.com/maps/api/js?sensor=false&amp;key=xxxxxxx"></script>
<!-- include plugin file-->
<script id="gmap" src="PATH_TO_YOUR_SCRIPTS/googlemapstyler.min.js"></script>
```

4. Set scripts  
With Vanilla Javascript
```
<script>
	window.onload=function(){
		var gmap = document.getElelemtsByClassName("gmap")
		var gmapstyler = new gmapStyler(gmap,{
			//options
			latitude:0,
			longitude:0,
			style:[
			  {
			    "elementType": "labels",
			    "stylers": [
			      { "visibility": "off" }
			    ]
			  },{
			    "stylers": [
			      { "saturation": -100 },
			      { "lightness": 48 }
			    ]
			  }
			]
		});

	}
</script>
```

With jQuery
```
<script>
	$(window).load(function(){
		$(".gmap").gmapStyler({
			//options
			// inner:".gmap-inner",
			latitude:0,
			longitude:0,
			style:[
			  {
			    "elementType": "labels",
			    "stylers": [
			      { "visibility": "off" }
			    ]
			  },{
			    "stylers": [
			      { "saturation": -100 },
			      { "lightness": 48 }
			    ]
			  }
			]
		});
	});
</script>
```

## Options
| option | type | description	 |
|----------|-----|-----|
| latitude | `number` | **required** latitude for center position	 |
| longitude | `number` | **required** longitude for center position	 |
| wrapper | `str` | HTML *class* for map wrapper (default : ```gmap```)	 |
| inner | `str` | HTML *class* for map inner(default : ```gmap-inner```)	 |
| style | `array` | an array containing styling google map	 |
| responsive | `boolean` | responsive map width/height (default : mono style) |
| zoom | `number` | zoom level (default : 17)   |
| pin | `str` | pin image path    |
| pinWidth | `number` | pin image width  |
| pinHeight | `number` | marker image height  |
| pinWidthSp | `number` | marker image width on smartphone |
| pinHeightSp | `number` | marker image height on smartphone |
 



CHANGELOG.md
<a name="1.0.0"></a>
# 1.0.0 (2017-08-31)





gulpfile.js
'use strict';

var gulp = require('gulp');
var uglify = require('gulp-uglify');
var plumber = require('gulp-plumber');
var notify = require('gulp-notify');
var Notifier = require('node-notifier');
var $ = require('gulp-load-plugins')({
  pattern: [
    'gulp-*',
    'gulp-notify',
    'browser-sync',
    'run-sequence',
    'gulp-plumber'
  ]
});

var config = {
  src: 'src/',
  dist: 'demo/',
  browserSupport: ['last 3 versions', 'ie >= 9', "ios 4", 'android >= 4']
};

gulp.task('js:dev', function() {
  return gulp.src([config.src +'js/config.js',config.src +'js/main.js'])
  .pipe(plumber({
            errorHandler: function(error){
             notify.onError("Error: <%= error.message %>")
             console.log(error.message); 
            }
  }))
  .pipe($.concat('googlemapstyler.js'))
  .pipe(gulp.dest(config.dist + 'js/'))
  .pipe(uglify())
  .pipe($.concat('googlemapstyler.min.js'))
  .pipe(gulp.dest(config.dist + 'js/'))
});


gulp.task('scss:dev', function() {
  return gulp.src(config.src + 'scss/{**/,}*.scss')
    .pipe(plumber({errorHandler: notify.onError('<%= error.message %>')}))
    .pipe($.sass({
      outputStyle: 'compressed',
      errLogToConsole: true
    }))
    .pipe($.autoprefixer({
      browsers: config.browserSupport
    }))
    .pipe(gulp.dest(config.dist + 'css/'))
    .pipe($.browserSync.stream())
});

gulp.task('watch:dev', function() {
 gulp.watch(
    config.src + 'scss/{**/,}*.scss',
    ['scss:dev']
  );
  gulp.watch(
    config.src + 'js/{**/,}*.js',
    ['js:dev']
  );
  gulp.watch([config.dist + '{**/,}*.html'])
    .on("change", $.browserSync.reload);
});

gulp.task('default', function() {
  $.runSequence(
    'watch:dev'
  );  
});



src/js/main.js
// -------------------------- styler definition  -------------------------- //

var gmapStyler = function gmapStyler(element, options) {

    var self = this;
    var _element = element;
    var _options = options;
    var _jQ = (!_options ? true : false);
    var _target = (_jQ ? this[0] : element);
    var _options = (_jQ ? _element : _options);
    _options = extend({}, _options, defaults);
    _options.latitude = _target.getAttribute("data-latitude");
    _options.longitude = _target.getAttribute("data-longitude");

    if (!_options.latitude || !_options.longitude) {
        throw new Error("latitude or longitude are not defined");
        return false;
    } else if (!_target) {
        throw new Error("Map HTML element does not exist");
        return false;
    }

    var _coord = new google.maps.LatLng(_options.latitude - 0.000000, _options.longitude);
    _options.settings.center = _coord;

    var _GMap = new google.maps.Map(_target.children[0], _options.settings);
    var _pin = (isMobile() ? {
        url: _options.pin,
        scaledSize: new google.maps.Size(_options.pinWidthSp, _options.pinHeightSp)
    } : _options.pin);
    var _GMarker = new google.maps.Marker({
        position: _coord,
        map: _GMap,
        icon: _pin
    });

    _GMap.setOptions({
        styles: _options.style
    });


    if(_options.resizeable || _options.responsive){
        google.maps.event.addDomListener(window, 'resize', function() {
          _GMap.setCenter(_coord);
      });
    }


    _target.addEventListener("mouseover", function() {
        _GMap.setOptions({
            styles: _options.hoverStyle
        });
    });
    _target.addEventListener("mouseleave", function() {
        _GMap.setOptions({
            styles: _options.hoverStyle
        });
    });

    if (isMobile()) {
        _target.addEventListener("touchstart", function() {
            _GMap.setOptions({
                styles: _options.hoverStyle
            });
        });
        _target.addEventListener("touchend", function() {
            _GMap.setOptions({
                styles: _options.style
            });
        });
    }

}


var extend = function(out) {
    out = out || {};

    for (var i = 1; i < arguments.length; i++) {
        if (!arguments[i])
            continue;

        for (var key in arguments[i]) {
            if (arguments[i].hasOwnProperty(key))
                out[key] = arguments[i][key];
        }
    }

    return out;
};

if (window.jQuery) {
    $.fn.gmapStyler = gmapStyler;
}

function isMobile() {
    if ($("body").hasClass('ua-sp')) {
        return true;
    } else {
        return false;
    }
}




src/js/config.js
/*!
 * Google map styler v1.0
 * Styling google map library
 * https://github.com/fs-nishimura/google-map-styler
 * MIT License
 * by fs-nishimura
 */


"use strict";


// -------------------------- default  -------------------------- //
var gMapOptions = {
  zoom: 17,
  mapTypeId: google.maps.MapTypeId.ROADMAP,
  scrollwheel: false,
  zoomControlOptions: {
  style: google.maps.ZoomControlStyle.SMALL
  },
  panControl: false,
  mapTypeControl: false,
  scaleControl: false,
  streetViewControl: false,
  rotateControl: false
};
var gMapStyleArrOff = [
  {
    "elementType": "labels",
    "stylers": [
      { "visibility": "off" }
    ]
  },{
    "stylers": [
      { "saturation": -100 },
      { "lightness": 48 }
    ]
  }
];
var gMapStyleArr = [
  {
    "elementType": "labels",
    "stylers": [
      { "visibility": "on" }
    ]
  },{
    "stylers": [
      { "saturation": -100 },
      { "lightness": 48 }
    ]
  }
];

var defaults = {
    latitude:1,
    longitude:1,
    style:gMapStyleArrOff,
    hoverStyle:gMapStyleArr,
    settings:gMapOptions,
    responsive:false,
    resizeable:true,
    pin:"./img/pin.png",
    pinWidth:20,
    pinHeight:20,
    pinWidthSp:10,
    pinHeightSp:10
  };


demo/js/googlemapstyler.js
/*!
 * Google map styler v1.0
 * Styling google map library
 * https://github.com/fs-nishimura/google-map-styler
 * MIT License
 * by fs-nishimura
 */


"use strict";


// -------------------------- default  -------------------------- //
var gMapOptions = {
  zoom: 17,
  mapTypeId: google.maps.MapTypeId.ROADMAP,
  scrollwheel: false,
  zoomControlOptions: {
  style: google.maps.ZoomControlStyle.SMALL
  },
  panControl: false,
  mapTypeControl: false,
  scaleControl: false,
  streetViewControl: false,
  rotateControl: false
};
var gMapStyleArrOff = [
  {
    "elementType": "labels",
    "stylers": [
      { "visibility": "off" }
    ]
  },{
    "stylers": [
      { "saturation": -100 },
      { "lightness": 48 }
    ]
  }
];
var gMapStyleArr = [
  {
    "elementType": "labels",
    "stylers": [
      { "visibility": "on" }
    ]
  },{
    "stylers": [
      { "saturation": -100 },
      { "lightness": 48 }
    ]
  }
];

var defaults = {
    latitude:1,
    longitude:1,
    style:gMapStyleArrOff,
    hoverStyle:gMapStyleArr,
    settings:gMapOptions,
    responsive:false,
    resizeable:true,
    pin:"./img/pin.png",
    pinWidth:20,
    pinHeight:20,
    pinWidthSp:10,
    pinHeightSp:10
  };

// -------------------------- styler definition  -------------------------- //

var gmapStyler = function gmapStyler(element, options) {

    var self = this;
    var _element = element;
    var _options = options;
    var _jQ = (!_options ? true : false);
    var _target = (_jQ ? this[0] : element);
    var _options = (_jQ ? _element : _options);
    _options = extend({}, _options, defaults);
    _options.latitude = _target.getAttribute("data-latitude");
    _options.longitude = _target.getAttribute("data-longitude");

    if (!_options.latitude || !_options.longitude) {
        throw new Error("latitude or longitude are not defined");
        return false;
    } else if (!_target) {
        throw new Error("Map HTML element does not exist");
        return false;
    }

    var _coord = new google.maps.LatLng(_options.latitude - 0.000000, _options.longitude);
    _options.settings.center = _coord;

    var _GMap = new google.maps.Map(_target.children[0], _options.settings);
    var _pin = (isMobile() ? {
        url: _options.pin,
        scaledSize: new google.maps.Size(_options.pinWidthSp, _options.pinHeightSp)
    } : _options.pin);
    var _GMarker = new google.maps.Marker({
        position: _coord,
        map: _GMap,
        icon: _pin
    });

    _GMap.setOptions({
        styles: _options.style
    });


    if(_options.resizeable || _options.responsive){
        google.maps.event.addDomListener(window, 'resize', function() {
          _GMap.setCenter(_coord);
      });
    }


    _target.addEventListener("mouseover", function() {
        _GMap.setOptions({
            styles: _options.hoverStyle
        });
    });
    _target.addEventListener("mouseleave", function() {
        _GMap.setOptions({
            styles: _options.hoverStyle
        });
    });

    if (isMobile()) {
        _target.addEventListener("touchstart", function() {
            _GMap.setOptions({
                styles: _options.hoverStyle
            });
        });
        _target.addEventListener("touchend", function() {
            _GMap.setOptions({
                styles: _options.style
            });
        });
    }

}


var extend = function(out) {
    out = out || {};

    for (var i = 1; i < arguments.length; i++) {
        if (!arguments[i])
            continue;

        for (var key in arguments[i]) {
            if (arguments[i].hasOwnProperty(key))
                out[key] = arguments[i][key];
        }
    }

    return out;
};

if (window.jQuery) {
    $.fn.gmapStyler = gmapStyler;
}

function isMobile() {
    if ($("body").hasClass('ua-sp')) {
        return true;
    } else {
        return false;
    }
}


